### DISCUSSION TITLE
Introduce (optional) values for tags

### DISCUSSION THREAD

**Discussion by jotaen** [jotaen/klog#358] [Feature ideas & discussions]
I think it would be useful to introduce a syntax for tags to have (optional) values assigned to them. For example:

```
1999-05-31
    4h Worked on #ticket=462
```

The rules could be this:
- A tag may be followed by an equals sign (`=`) plus a value.
  - In the above example, the tag itself is still `#ticket`. In terms of terminology, I’d suggest `ticket` to be the tag’s name and `462` to be the tag’s value. “Tag” refers to the entire concept.
  - The allowed characters for values are the same as for tags themselves: letters, digits, underscores [and dashes](https://github.com/jotaen/klog/issues/168). In contrast to the tag name, however, the value is **not** harmonised to lower-case. So `#tag=VALUE` ­≠ `#tag=value`, whereas `#TAG=value` = `#tag=value`.
  - The value could optionally be quoted, e.g. `#ticket="462"` or `#ticket='462'`. That would allow values to contain more characters, e.g. `#ticket="Foo Bar #.:!"` → `Foo Bar #.:!`, whereas without quotes it’d be `#ticket=Foo Bar #.:!` → `Foo`, because the space character terminates the tag.
    - The quotes themselves are not part of the values.
    - If the closing quote is missing, the scope of the value ends at the end of the line. (So multiline values are prohibited.)
    - Escaping of quotes wouldn’t be supported, because that would be a bit too technical for klog. So no `#ticket='It\'s done'`.
- Absent value (`#ticket` or `#ticket=` or `#ticket=""`) is treated as empty string. That way, tags without values behave exactly as they do right now.
- The `=` can only appear once, so `#ticket=1=2` is either invalid or treated as `#ticket=1`.

On the CLI, users could additionally query for tag values. So with the example above:

- `klog total --tag ticket` would return `4h`
- `klog total --tag ticket=462` would return `4h`
- `klog total --tag ticket=1` would return `0h`

And `klog tags` could additionally break it down by value as well:

```
#code-reviews  7h
#ticket        18h45m
 =465          6h30m
 =689          7h
 =911          5h15m
#conference    24h
```

## Discussion
- An alternative to the `=` would be a colon, so `#ticket:462`. I think the meaning is a bit clearer with an equals sign, however, and it would also avoid ambiguity with a colon that appears within the sentence, like `Worked on a #ticket: it’s not done yet, though.` Therefore, I think the `=` is the better option.
- The quoting introduces most syntactical complexity. I think it might still be valuable, though, because letters and digits could be limited for some use cases. The default way would be without quotes also.
- If quoting is allowed for values, one could make a point to also allow that for tag names, e.g. `#"foo.bar"`. That looks awkward to me, however, and you could as well use a value then, e.g. `#whatever="foo.bar"`, so I wouldn’t do it.

This is a breaking change, because the `=` currently terminates the tag. The meaning would change now.

**Comment by vladdeSV** [comment]
Just wanted to mention that I really like this idea. The decision to use `=` is a good one.

**Comment by jotaen** [comment]
Resolved via https://github.com/jotaen/klog/pull/181.
